drivers/gpio/gpio_mcp23s17.c

```diff
@@ -0,0 +1,515 @@
+		LOG_DBG("spi_transceive FAIL %d\n", ret);
+		return ret;
+	}
+
+	LOG_DBG("MCP23S17: Read: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X",
+		reg, buf->byte[0], (reg + 1), buf->byte[1]);
+
+	return 0;
+}
+
+/**
+ * @brief Write both port 0 and port 1 registers of certain register function.
+ *
+ * Given the register in reg, write the pair of port 0 and port 1.
+ *
+ * @param dev Device struct of the MCP23S17.
+ * @param reg Register to write into (the PORTA of the pair of registers).
+ * @param buf Buffer to write data from.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int write_port_regs(struct device *dev, u8_t reg,
+			    union mcp23s17_port_data *buf)
+{
+	struct mcp23s17_drv_data *const drv_data =
+	    (struct mcp23s17_drv_data * const)dev->driver_data;
+	int ret;
+
+	u8_t addr = MCP23S17_ADDR;
+	u8_t buffer_tx[2] = { addr, reg };
+
+	const struct spi_buf tx_buf[2] = {
+		{
+			.buf = buffer_tx,
+			.len = 2,
+		},
+		{
+			.buf = &buf->byte,
+			.len = 2,
+		}
+	};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = 2,
+	};
+
+	ret = spi_write(drv_data->spi, &drv_data->spi_cfg, &tx);
+	if (ret) {
+		LOG_DBG("spi_write FAIL %d\n", ret);
+		return ret;
+	}
+
+	LOG_DBG("MCP23S17: Write: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X",
+		reg, buf->byte[0], (reg + 1), buf->byte[1]);
+	return 0;
+}
+
+/**
+ * @brief Setup the pin direction (input or output)
+ *
+ * @param dev Device struct of the MCP23S17
+ * @param pin The pin number
+ * @param flags Flags of pin or port
+ *
+ * @return 0 if successful, failed otherwise
+ */
+static int setup_pin_dir(struct device *dev, u32_t pin, int flags)
+{
+	struct mcp23s17_drv_data *const drv_data =
+	    (struct mcp23s17_drv_data * const)dev->driver_data;
+	union mcp23s17_port_data *dir = &drv_data->reg_cache.iodir;
+	union mcp23s17_port_data *output = &drv_data->reg_cache.gpio;
+	int ret;
+
+	if ((flags & GPIO_OUTPUT) != 0U) {
+		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0U) {
+			output->all |= BIT(pin);
+		} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0U) {
+			output->all &= ~BIT(pin);
+		}
+		dir->all &= ~BIT(pin);
+	} else {
+		dir->all |= BIT(pin);
+	}
+
+	ret = write_port_regs(dev, REG_GPIO_PORTA, output);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = write_port_regs(dev, REG_IODIR_PORTA, dir);
+
+	return ret;
+}
+
+/**
+ * @brief Setup the pin pull up/pull down status
+ *
+ * @param dev Device struct of the MCP23S17
+ * @param pin The pin number
+ * @param flags Flags of pin or port
+ *
+ * @return 0 if successful, failed otherwise
+ */
+static int setup_pin_pullupdown(struct device *dev, u32_t pin, int flags)
+{
+	struct mcp23s17_drv_data *const drv_data =
+	    (struct mcp23s17_drv_data * const)dev->driver_data;
+	union mcp23s17_port_data *port;
+	u16_t bit_mask;
+	u16_t new_value = 0;
+	int ret;
+
+	/* Setup pin pull up or pull down */
+	port = &drv_data->reg_cache.gppu;
+	bit_mask = 1 << pin;
+
+	/* pull down == 0, pull up == 1 */
+	if ((flags & GPIO_PUD_MASK) == GPIO_PULL_UP) {
```

Comment:
Unfortunately the current GPIO API doesn't include deprecated pragma (only text description) so it's easy to miss it, however all `GPIO_PUD_` macros are depracated and will be later removed. It's best to directly test the bit as in 
```
	if ((flags & GPIO_PULL_UP) != 0) {
```